### ISSUE TITLE
Spacing is the same for each axis in the lowres configuration despite large resolution differences.

### ISSUE THREAD

**Issue by karol-szustakowski** [MIC-DKFZ/nnUNet#2764]
Hello, and thank you for this amazing work.
According to your paper:

> The input patch size of the 3D lowres U-Net must cover at least 1/4 of
> the median shape of the resampled trainig cases to ensure sufficient contextual information. Higher
> resolution axes are downsampled first, resulting in a potentially different aspect ratio of the data
> relative to the full resolution data.

I am currently working with images with spatial resolution of 1 unit/voxel on the x  and y axis, and 3 units/voxel on the z axis.
This is reflected in the plans json:
```json
    "original_median_spacing_after_transp": [
        3.119999885559082,
        1.0,
        1.0
    ],
```
However, after preprocessing, this is the config I get:
```json
        "3d_lowres": {
            ...,
            "patch_size": [
                40,
                256,
                224
            ],
            "median_image_size_in_voxels": [
                62,
                371,
                371
            ],
            "spacing": [
                5.391651440886649,
                5.391651440886649,
                5.391651440886649
            ],
            "normalization_schemes": [
                "ZScoreNormalization"
            ],
            "use_mask_for_norm": [
                false
            ],
            ...,
            "resampling_fn_data_kwargs": {
                "is_seg": false,
                "order": 3,
                "order_z": 0,
                "force_separate_z": null
            },
            "resampling_fn_seg_kwargs": {
                "is_seg": true,
                "order": 1,
                "order_z": 0,
                "force_separate_z": null
            },
            "resampling_fn_probabilities": "resample_data_or_seg_to_shape",
            "resampling_fn_probabilities_kwargs": {
                "is_seg": false,
                "order": 1,
                "order_z": 0,
                "force_separate_z": null
            },
            "architecture": {
                "network_class_name": "dynamic_network_architectures.architectures.unet.PlainConvUNet",
                "arch_kwargs": {
                    "n_stages": 6,
                    "features_per_stage": [
                        32,
                        64,
                        128,
                        256,
                        320,
                        320
                    ],
                    "conv_op": "torch.nn.modules.conv.Conv3d",
                    "kernel_sizes": [
                        [
                            3,
                            3,
                            3
                        ],
                        [
                            3,
                            3,
                            3
                        ],
                        [
                            3,
                            3,
                            3
                        ],
                        [
                            3,
                            3,
                            3
                        ],
                        [
                            3,
                            3,
                            3
                        ],
                        [
                            3,
                            3,
                            3
                        ]
                    ],
                    "strides": [
                        [
                            1,
                            1,
                            1
                        ],
                        [
                            2,
                            2,
                            2
                        ],
                        [
                            2,
                            2,
                            2
                        ],
                        [
                            2,
                            2,
                            2
                        ],
                        [
                            1,
                            2,
                            2
                        ],
                        [
                            1,
                            2,
                            2
                        ]
                    ],
                    "n_conv_per_stage": [
                        2,
                        2,
                        2,
                        2,
                        2,
                        2
                    ],
                    ...,
            },
            "batch_dice": false,
            "next_stage": "3d_cascade_fullres"
        },
```

For me, it is very important to keep all the possible details from the axis with the lower resolution, as they will definitely alter the result, but from what I see, the spacing remains the same for every axis, and the only thing that might help are kernels with stride (1, 2, 2). Is there something I am missing?

**Comment by seziegler** [comment]
Hi, is the issue still relevant?

**Comment by karol-szustakowski** [comment]
Not for me personally, but I believe that this might be an interesting option to include for others as well.

**Comment by seziegler** [comment]
The 3d_lowres configuration is not meant to preserve the original anisotropy ratio of the data. Its purpose is to increase field of view so that the lowres patch covers enough of the median case for the cascade. The planner therefore starts from the fullres spacing and increases the target spacing until the patch-size coverage criterion is met; it first adjusts the anisotropic axis until the ratio is within about a factor of 2, then scales all axes together. That is why 3d_lowres may become nearly isotropic or fully isotropic.

The fine-grained information is expected to be recovered in 3d_cascade_fullres, while the lowres stage mainly provides global context. The (1, 2, 2) strides you observed are also part of the mechanism that avoids downsampling the low-resolution axis too aggressively early on.

So: no, you are not missing anything. If you specifically want the lowres stage to keep a stronger anisotropy, you would need to create a custom configuration with manually overridden target spacing and rerun preprocessing for that config.
